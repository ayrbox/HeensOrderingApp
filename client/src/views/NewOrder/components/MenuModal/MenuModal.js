import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
// import DialogTitle from '@material-ui/core/DialogTitle';
// import TextField from '@material-ui/core/TextField';
// import FormControl from '@material-ui/core/FormControl';
// import FormHelperText from '@material-ui/core/FormHelperText';
// import CircularProgress from '@material-ui/core/CircularProgress';
// import Select from '@material-ui/core/Select';
// import InputLabel from '@material-ui/core/InputLabel';
// import Input from '@material-ui/core/Input';
// import MenuItem from '@material-ui/core/MenuItem';
// import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
// import AddIcon from '@material-ui/icons/Add';
// import DeleteIcon from '@material-ui/icons/Delete';
// import IconButton from '@material-ui/core/IconButton';
import Grid from '@material-ui/core/Grid';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import ListItemText from '@material-ui/core/ListItemText';
import Checkbox from '@material-ui/core/Checkbox';
import IconButton from '@material-ui/core/IconButton';
import CommentIcon from '@material-ui/icons/Comment';



const MenuModal = ({
  menu,
  open,
  resetMenu,
}) => (
  <Dialog
    open={open}
    aria-labelledby="Order Menu Modal"
    scroll="body"
    onClose={resetMenu}
  >
    <DialogContent>
      {menu && (
        <Fragment>
          <Grid
            container
            direction="row"
            alignItems="center"
            justify="space-between"
            spacing={8}
          >
            <Grid item>
              <Typography variant="h3">
                {menu.name}
              </Typography>
              <Typography variant="body1">
                {menu.description}
              </Typography>
            </Grid>
            <Grid item>
              <Typography variant="h1">
                Â£
                {menu.price.toFixed(2)}
              </Typography>
            </Grid>
          </Grid>
          <List>
            {
              menu.menuOptions.map(({ description, additionalCost }) => (
                <ListItem role={undefined} button>
                  <Checkbox
                    tabIndex={-1}
                    disableRipple
                  />
                  <ListItemText primary={description} />
                  { additionalCost && (
                    <h1>{additionalCost}</h1>
                  )}
                </ListItem>
              ))
            }
          </List>
        </Fragment>
      )}

      <pre>
        {JSON.stringify(menu, null, 2)}
      </pre>
    </DialogContent>
    <DialogActions>
      <Button
        onClick={resetMenu}
        color="primary"
      >
        Cancel
      </Button>
      <Button
        type="submit"
        color="primary"
      >
        Save
      </Button>
    </DialogActions>
  </Dialog>
);

MenuModal.propTypes = {
  menu: PropTypes.shape().isRequired,
  open: PropTypes.bool.isRequired,
  resetMenu: PropTypes.func.isRequired,
};

export default withStyles({})(MenuModal);
