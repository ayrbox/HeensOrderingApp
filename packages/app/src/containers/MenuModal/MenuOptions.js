import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Checkbox from '@material-ui/core/Checkbox';

const styles = {
  checkBox: {
    padding: 0,
  },
  optionCost: {
    width: '200px',
    textAlign: 'right',
  },
};

const MenuOptions = ({
  classes,
  options,
  onToggleSelect,
}) => (
  <List data-testid="menu-option-list">
    {options.map(({
      id,
      description,
      additionalCost,
      isChecked,
    }) => (
      <ListItem
        key={id}
        role="button"
        dense
        button
        onClick={onToggleSelect(id)}
        data-testid="menu-option-item"
      >
        <Checkbox
          tabIndex={-1}
          disableRipple
          className={classes.checkBox}
          checked={isChecked}
        />
        <ListItemText primary={description} />
        <ListItemText
          primary={additionalCost === 0 ? '' : `Â£${additionalCost.toFixed(2)}`}
          className={classes.optionCost}
        />
      </ListItem>
    ))}
  </List>
);

MenuOptions.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  classes: PropTypes.shape().isRequired,
  onToggleSelect: PropTypes.func.isRequired,
};

export default withStyles(styles)(MenuOptions);
