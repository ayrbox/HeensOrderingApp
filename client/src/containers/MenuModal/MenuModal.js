import React, { Fragment, useState, useEffect } from 'react';
import { pick } from 'lodash';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import Divider from '@material-ui/core/Divider';

import MenuOptions from './MenuOptions';

const toogleOptionChecked = (id, items) => {
  const idx = items.findIndex(({ id: _ }) => _ === id);

  if (idx === -1) return items;

  const item = Object.assign({}, items[idx], {
    isChecked: !items[idx].isChecked,
  });

  return [
    ...items.slice(0, idx),
    item,
    ...items.slice(idx + 1),
  ];
};

const getTotal = ({ price }, options) => options
  .filter(({ isChecked: _ }) => _)
  .reduce((total, { additionalCost: _ }) => total + _, price);

const renderTotal = (menu, options) => !!options.length && (
  <Fragment>
    <Divider />
    <div style={{ padding: '8px', textAlign: 'right' }}>
      <Typography variant="h1">
        £
        {getTotal(menu, options).toFixed(2)}
      </Typography>
    </div>
  </Fragment>
);

const getOrderItem = (menu, options) => ({
  ...pick(menu, ['name', 'description', 'price']),
  menuOptions: options.filter(({ isChecked: _ }) => _)
    .map(o => pick(o, ['description', 'additionalCost'])),
  itemTotal: getTotal(menu, options),
});

const MenuModal = ({
  menu,
  open,
  resetMenu,
  selectOrderItem,
}) => {
  const [options, setOptions] = useState([]);
  useEffect(() => {
    if (menu) {
      setOptions(menu.menuOptions.map(({
        _id: id,
        additionalCost,
        description,
      }) => ({
        id,
        additionalCost,
        description,
        isChecked: false,
      })));
    }
  }, [menu]);

  return (
    <Dialog
      open={open}
      aria-labelledby="Order Menu Modal"
      scroll="body"
      onClose={resetMenu}
    >
      <DialogContent>
        {menu && (
          <Fragment>
            <Grid
              container
              direction="row"
              alignItems="center"
              justify="space-between"
              spacing={8}
            >
              <Grid item>
                <Typography variant="h3">
                  {menu.name}
                </Typography>
                <Typography variant="body1">
                  {menu.description}
                </Typography>
              </Grid>
              <Grid item>
                <Typography variant="h1">
                  £
                  {menu.price.toFixed(2)}
                </Typography>
              </Grid>
            </Grid>
            <MenuOptions
              options={options}
              onToggleSelect={id => () => {
                setOptions(prev => toogleOptionChecked(id, prev));
              }}
            />
            {renderTotal(menu, options)}
          </Fragment>
        )}
      </DialogContent>
      <DialogActions>
        <Button
          onClick={resetMenu}
          color="primary"
        >
          Cancel
        </Button>
        <Button
          type="submit"
          color="primary"
          onClick={(e) => {
            e.preventDefault();
            const orderItem = getOrderItem(menu, options);
            selectOrderItem(orderItem);
          }}
        >
          Select
        </Button>
      </DialogActions>
    </Dialog>
  );
};

MenuModal.propTypes = {
  menu: PropTypes.shape().isRequired,
  open: PropTypes.bool.isRequired,
  resetMenu: PropTypes.func.isRequired,
  selectOrderItem: PropTypes.func.isRequired,
};

export default withStyles({})(MenuModal);
