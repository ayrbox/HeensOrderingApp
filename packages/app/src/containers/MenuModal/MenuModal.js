import React, { Fragment, useState, useEffect } from 'react';
import { pick } from 'lodash';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import Divider from '@material-ui/core/Divider';

import MenuOptions from './MenuOptions';

const toogleOptionChecked = (id, items) => {
  const idx = items.findIndex(({ id: _ }) => _ === id);

  if (idx === -1) return items;

  const item = Object.assign({}, items[idx], {
    isChecked: !items[idx].isChecked,
  });

  return [
    ...items.slice(0, idx),
    item,
    ...items.slice(idx + 1),
  ];
};

const getTotal = ({ price }, options) => options
  .filter(({ isChecked: _ }) => _)
  .reduce((total, { additionalCost: _ }) => total + _, price);

const renderTotal = (menu, options) => !!options.length && (
  <Fragment>
    <Divider />
    <div style={{ padding: '8px', textAlign: 'right' }}>
      <Typography variant="h1">
        £
        {getTotal(menu, options).toFixed(2)}
      </Typography>
    </div>
  </Fragment>
);

const getOrderItem = (menu, options) => ({
  ...pick(menu, ['name', 'description', 'price']),
  menuOptions: options.filter(({ isChecked: _ }) => _)
    .map(o => pick(o, ['description', 'additionalCost'])),
  itemTotal: getTotal(menu, options),
});

const MenuModal = ({
  menu,
  open,
  resetOrder,
  addOrderItem,
}) => {
  const [options, setOptions] = useState([]);
  useEffect(() => {
    if (menu) {
      setOptions(menu.menuOptions.map(({
        _id: id,
        additionalCost,
        description,
      }) => ({
        id,
        additionalCost,
        description,
        isChecked: false,
      })));
    }
  }, [menu]);

  return (
    <Dialog
      open={open}
      aria-labelledby="Order Menu Modal"
      scroll="body"
      onClose={resetOrder}
      data-testid="menu-dialog"
    >
      <DialogContent>
        {menu && (
          <Fragment>
            <Grid
              container
              direction="row"
              alignItems="center"
              justify="space-between"
              spacing={8}
            >
              <Grid item>
                <Typography
                  variant="h3"
                  data-testid="menu-name"
                >
                  {menu.name}
                </Typography>
                <Typography
                  variant="body1"
                  data-testid="menu-description"
                >
                  {menu.description}
                </Typography>
              </Grid>
              <Grid item>
                <Typography
                  variant="h1"
                  data-testid="menu-price"
                >
                  £
                  {menu.price.toFixed(2)}
                </Typography>
              </Grid>
            </Grid>
            <MenuOptions
              options={options}
              onToggleSelect={id => () => {
                setOptions(prev => toogleOptionChecked(id, prev));
              }}
            />
            {renderTotal(menu, options)}
          </Fragment>
        )}
      </DialogContent>
      <DialogActions>
        <Button
          onClick={resetOrder}
          color="primary"
          data-testid="button-cancel"
        >
          Cancel
        </Button>
        <Button
          type="submit"
          color="primary"
          onClick={(e) => {
            e.preventDefault();
            const orderItem = getOrderItem(menu, options);
            addOrderItem(orderItem);
          }}
          data-testid="button-select"
        >
          Select
        </Button>
      </DialogActions>
    </Dialog>
  );
};

MenuModal.propTypes = {
  menu: PropTypes.shape().isRequired,
  open: PropTypes.bool.isRequired,
  resetOrder: PropTypes.func.isRequired,
  addOrderItem: PropTypes.func.isRequired,
};

export default withStyles({})(MenuModal);
