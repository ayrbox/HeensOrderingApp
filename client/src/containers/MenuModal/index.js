import { connect } from 'react-redux';
import MenuModal from './MenuModal';

// import { resetMenu, selectOrderItem } from '../../actions/newOrderActions';

const mapState = state => ({
  menu: state.newOrder.menu,
  open: state.newOrder.openMenu,
});

export default connect(mapState, {
  resetMenu: () => {},
  selectOrderItem: () => {},
})(MenuModal);
