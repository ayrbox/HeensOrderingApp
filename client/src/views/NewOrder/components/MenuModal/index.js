import { connect } from 'react-redux';
import MenuModal from './MenuModal';

import { resetMenu, selectOrderItem } from '../../../../actions/orderActions';

const mapState = state => ({
  menu: state.orders.menu,
  open: state.orders.openMenu,
});

export default connect(mapState, {
  resetMenu,
  selectOrderItem,
})(MenuModal);
