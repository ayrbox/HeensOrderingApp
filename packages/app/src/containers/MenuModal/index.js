import { connect } from 'react-redux';
import MenuModal from './MenuModal';

import { resetOrder, addOrderItem } from '../../store/newOrder';

const mapState = state => ({
  menu: state.newOrder.selectedMenu,
  open: state.newOrder.openMenuModal,
});

export default connect(mapState, {
  resetOrder,
  addOrderItem,
})(MenuModal);
