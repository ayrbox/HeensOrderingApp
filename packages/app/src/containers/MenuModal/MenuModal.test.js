import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import MenuModal from './MenuModal';

const mockAddItem = jest.fn();
const mockReset = jest.fn();

const sampleMenu = {
  tags: ['Wheat', 'Crustaceans', 'Egg', 'Fish', 'Peanuts', 'Sesame'],
  _id: '5d631866c5799434e5c1ac4d',
  name: 'Heen\'s Mixed Hors D\'oeuvreså¹¶',
  description: 'Mixed Hors D\'oeuvres (For 2 Persons or More)',
  price: 7.9,
  category: {
    _id: '5b3dfd4116987b4e407f8652',
    name: 'Appetisers',
    description: 'Appetisers Menu',
  },
  menuOptions: [],
};

describe('<MenuModal />', () => {
  describe('when open is false', () => {
    it('should not renders dialog', () => {
      const { container, queryByTestId } = render(
        <MenuModal
          addOrderItem={mockAddItem}
          resetOrder={mockReset}
          open={false}
          menu={sampleMenu}
        />,
      );
      expect(queryByTestId('menu-dialog')).toBeNull();
      expect(container).toMatchSnapshot();
    });
  });

  describe('when open is true', () => {
    it('should render dialog correctly', () => {
      const { container, getByTestId } = render(
        <MenuModal
          addOrderItem={mockAddItem}
          resetOrder={mockReset}
          open
          menu={sampleMenu}
        />,
      );
      getByTestId('menu-dialog');
      expect(getByTestId('menu-name')).toHaveTextContent(sampleMenu.name);
      expect(getByTestId('menu-description')).toHaveTextContent(sampleMenu.description);
      expect(getByTestId('menu-price')).toHaveTextContent(sampleMenu.price);

      fireEvent.click(getByTestId(/button-select/));
      expect(mockAddItem).toHaveBeenCalled(); // can be assert with `toHaveBeenCalledWith`

      fireEvent.click(getByTestId(/button-cancel/));
      expect(mockReset).toHaveBeenCalled();

      expect(container).toMatchSnapshot();
    });
  });
});
