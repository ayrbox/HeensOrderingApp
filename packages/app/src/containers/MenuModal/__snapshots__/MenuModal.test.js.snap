// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuModal /> renders correctly 1`] = `<div />`;

exports[`<MenuModal /> when open is false should not renders dialog 1`] = `<div />`;

exports[`<MenuModal /> when open is true should render dialog correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
